crates/project-model/src/workspace.rs

```diff
@@ -1901,3 +1899,23 @@ fn sysroot_metadata_config(
         kind: "sysroot",
     }
 }
+
+fn cargo_target_dir(
+    manifest: &ManifestPath,
+    extra_env: &FxHashMap<String, Option<String>>,
+    sysroot: &Sysroot,
+) -> Option<Utf8PathBuf> {
+    let cargo = sysroot.tool(Tool::Cargo, manifest.parent(), extra_env);
+    let mut meta = cargo_metadata::MetadataCommand::new();
+    meta.cargo_path(cargo.get_program());
+    meta.manifest_path(manifest);
+    // `--no-deps` doesn't (over)write lockfiles as it doesn't do any package resolve.
+    // So we can use it to get `target_directory` before copying lockfiles
+    let mut other_options = vec!["--no-deps".to_owned()];
+    if manifest.is_rust_manifest() {
+        meta.env("RUSTC_BOOTSTRAP", "1");
+        other_options.push("-Zscript".to_owned());
+    }
+    meta.other_options(other_options);
+    meta.exec().map(|m| m.target_directory).ok()
+}
```

Comment:
We can be even smarter here. For one this code path will be exercised in the majority of cases. And note how we fallback to `--no-deps` when the command otherwise fails to semi support offline modes and the like. So I think it would make sense to just run this command proper (as we would usually) with `--no-deps`. If that succeeds, then run it again without while using the target dir info and stuff, and if this then fails, fallback to using the `--no-deps` version (this way we save one metadata invocation).